src/Tasks/Microsoft.NET.Build.Tasks/GetPackagesToPrune.cs

```diff
@@ -72,51 +83,208 @@ protected override void ExecuteCore()
+                        packagesForFrameworkReference = LoadPackagesToPruneFromTargetingPack(log, key.TargetFrameworkIdentifier, key.TargetFrameworkVersion, frameworkReference, targetingPackRoots);
+                    }
+
+                    //  Fall back to framework packages data for older framework for WindowsDesktop if necessary
+                    //  https://github.com/dotnet/windowsdesktop/issues/4904
+                    if (packagesForFrameworkReference == null && frameworkReference.Equals("Microsoft.WindowsDesktop.App", StringComparison.OrdinalIgnoreCase))
+                    {
+                        log.LogMessage("Failed to load prune package data for WindowsDesktop from targeting packs, loading from framework packages instead");
+                        packagesForFrameworkReference = LoadPackagesToPruneFromFrameworkPackages(key.TargetFrameworkIdentifier, key.TargetFrameworkVersion, frameworkReference,
+                            acceptNearestMatch: true);
+                    }
+                }
+
+                if (packagesForFrameworkReference == null)
+                {
+                    //  We didn't find the data for packages to prune.  This indicates that there's a bug in the SDK construction, so fail hard here so that we fix that
+                    //  (rather than a warning that might be missed).
+                    //  Since this indicates an error in the SDK build, the message probably doesn't need to be localized.
+
+                    if (allowMissingPrunePackageData)
+                    {
+                        log.LogMessage($"Prune package data not found for {key.TargetFrameworkIdentifier} {key.TargetFrameworkVersion} {frameworkReference}");
+                    }
+                    else
+                    {
+                        log.LogError(Strings.PrunePackageDataNotFound, key.TargetFrameworkIdentifier, key.TargetFrameworkVersion, frameworkReference);
+                    }
+                }
+                else
+                {
+                    AddPackagesToPrune(packagesToPrune, packagesForFrameworkReference.Select(kvp => (kvp.Key, kvp.Value)), log);
                 }
             }
 
-            PackagesToPrune = packagesToPrune.Select(p =>
+            return packagesToPrune.Select(p =>
             {
                 var item = new TaskItem(p.Key);
-                item.SetMetadata("Version", p.Value.ToString());
+
+                //  If a given version of a package is included in a framework, assume that any patches
+                //  to that package will be included in patches to the framework, and thus should be pruned.
+                //  See https://github.com/dotnet/sdk/issues/44566
+                //  To do this, we set the patch version for the package to be pruned to 32767, which should be
+                //  higher than any actual patch version.
+                var maxPatch = new NuGetVersion(p.Value.Major, p.Value.Minor, 32767);
```

Comment:
I don't think we should do this for .NETFramework - since there we version the assembly on every package release.

Will this apply for .NETFramework?